h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -1479,39 +1479,79 @@ public static long convertScale(long nanosOfDay, int scale) {
      * @return date truncated to 'day'
      */
     public static Value truncateDate(String timeUnit, Value value) {
-        Value result = null;
+        Value result;
 
-        // Retrieve the dateValue.
+        // Retrieve the dateValue and the time in nanoseconds if the date.
         long[] fieldDateAndTime = DateTimeUtils.dateAndTimeFromValue(value);
         long dateValue = fieldDateAndTime[0];
+        long timeNanosRetrieved = fieldDateAndTime[1];
 
-        // Case where the date has to be truncated to the day.
-        if (timeUnit.equals("DAY")) {
+        // Variable that contains the number of nanoseconds in 'hour', 'minute',
+        // etc.
+        long nanoInHour = 3_600_000_000_000l;
+        long nanoInMinute = 60_000_000_000l;
+        long nanoInSecond = 1_000_000_000l;
+        long nanoInMilliSecond = 1_000_000l;
+        long nanoInMicroSecond = 1_000l;
```

Comment:
If you don't want to create constants for them (this may be not so necessary), just move them in corresponding code blocks. Otherwise interpreter will initialize them all on each invocation. JIT can optimize it, but compilation is performed only for hot methods.